### ISSUE TITLE
When not to use QUIC

### ISSUE THREAD

**Issue by tekjar** [djc/quinn#473]
Hi. I've been planning to understand the protocol and contribute. I came across [this](https://news.ycombinator.com/item?id=19476439)  post while digging. I have a few questions

* Are the numbers exaggerated as per your experience?
* What are the drawbacks of using QUIC?
* Is it suitable for IOT workloads where packet sizes are small (ranges from bytes to few kilo bytes) 
* Are any of the issues in the post solved in newer versions of the draft?

**Comment by djc** [comment]
Hi, thanks for your questions!

Honestly, as far as I know we haven't done any extensive benchmarks on Quinn or QUIC. So far we've found that Quinn performance is not bad when compared to other implementations, but of course none of these implementations are very mature compared to the TCP world. Also given that the protocol lives in user space, there might be wins in moving more of the packet handling into userspace as well using something like DPDK, but we don't have experience with that either. We also haven't done any tests in that direction, although we've taken care to separate our protocol logic in a separate crate so that it would be potentially not as hard to plugin different network layers.

We do know that our `Connection` type takes up a relatively large amount of memory (on the other of a kB) and that doesn't even count some of the buffers yet, so memory use might be an issue on some kinds of IoT.

As far as I know QUICv1 changes haven't really optimized for these sorts of use cases even in later drafts, but I'm not very closely plugged in to the reasoning behind protocol changes. @Ralith might know more here.

Bottom line: I don't think there are any easy answers here and you'd probably want to run some benchmarks specific to the scenario you have in mind to figure out if QUIC/Quinn will serve your needs. We are generally quite responsive to PRs and issues, so happy to support you if you decide to dig in a little. We're also usually responsive in our Gitter channel if more synchronous dialogue would help.

**Comment by tekjar** [comment]
> We are generally quite responsive to PRs and issues, so happy to support you if you decide to dig in a little

Awesome. I might be able to help with numbers w.r.t a real world use case. I'm implementing an mqtt broker which supports both tcp and quic

**Comment by Ralith** [comment]
> Are the numbers exaggerated as per your experience?

I think that post is incorrect to focus specifically on syscall counts, because it's easy to reduce those under high load (i.e. the only time it matters) using sendmmsg/recvmmsg. ACK rate can also be tuned in a number of ways. However, the overall point is correct: QUIC does require substantially more CPU than TCP. There are various efforts ongoing to mitigate this, ranging from [new kernel interfaces](http://vger.kernel.org/lpc_net2018_talks/willemdebruijn-lpc2018-udpgso-paper-DRAFT-1.pdf) and various types of hardware offload.

That said, currently attainable performance should be more than adequate for typical internet applications. Only in cases where extremely high rates data are being exchanged with very little application-level processing should transport performance be a serious bottleneck, fixable issues in a particular QUIC implementation aside.

> Is it suitable for IOT workloads where packet sizes are small (ranges from bytes to few kilo bytes)

QUIC has low bandwidth overhead in general, and is fine for low data rate applications. However, the cryptographic handshake involves a number of large packets that may pose an issue for extremely narrow (e.g. <1Kbps) links. QUIC is designed to have a pluggable cryptographic layer and Quinn exposes this, so in principle you could swap out the TLS embedding with something like the noise protocol to mitigate this. Let us know if you want to try that; there are some security implications and the interface may need refinement. Such a scheme would not strictly be standard QUIC, but it's a case we're interested in supporting.

 > Are any of the issues in the post solved in newer versions of the draft?

QUIC is not specifically trying to be as cheap as TCP; it is a more versatile and more secure protocol and that doesn't come for free. If you don't need to serve 90Gb/s with a 12-core Xeon-D, the difference in cost probably doesn't matter to you. Not all applications are in scope: QUIC will probably never be a protocol you want to use to ship massive amounts of data over a secure, isolated network, unless hardware offload comes a long way.

Lots of effort is being invested in finding ways to make QUIC implementations faster, the focus is in kernels, hardware, and implementation details rather than in the draft itself.

**Comment by tekjar** [comment]
Thanks for the answers @djc and @Ralith
